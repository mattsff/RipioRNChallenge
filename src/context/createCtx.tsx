import React from 'react';

// create context with no upfront defaultValue
// without having to do undefined check all the time
// https://github.com/typescript-cheatsheets/react-typescript-cheatsheet/blob/master/README.md#context
export function createCtx<A>(name: string) {
  const ctx = React.createContext<A | undefined>(undefined);
  ctx.displayName = name;
  function useCtx() {
    const c = React.useContext(ctx);
    if (!c) {
      throw new Error('useCtx must be inside a Provider with a value');
    }
    return c;
  }
  return [useCtx, ctx.Provider, ctx] as const; // make TypeScript infer a tuple, not an array of union types
}
