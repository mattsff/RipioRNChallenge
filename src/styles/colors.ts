export const Colors = {
  Transparent: '#00000000',
  Overlay: '#0000006D',
};
