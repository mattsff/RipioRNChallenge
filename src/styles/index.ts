export * from './typography';
export * from './spacing';
export * from './colors';
