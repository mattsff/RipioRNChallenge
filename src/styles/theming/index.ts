export { LightTheme } from './LightTheme';
export { DarkTheme } from './DarkTheme';
export type { ITheme } from './ITheme';
export { EThemeType } from './ITheme';
export { RNElementsTheme } from './RNElementsTheme';
