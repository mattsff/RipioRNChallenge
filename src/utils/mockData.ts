import { Currency } from '../types/currency';

export const bitcoin: Currency = {
  scale: 8,
  name: 'Bitcoin',
  ticker: 'BTC',
  icon: 'bitcoin',
  iconColor: '#F7931A',
  cashInEnabled: false,
  cashOutEnabled: true,
};


export const ars: Currency = {
  scale: 2,
  name: 'Pesos Argentinos',
  ticker: 'ARS',
  icon: '',
  iconColor: '',
  cashInEnabled: false,
  cashOutEnabled: false,
};

