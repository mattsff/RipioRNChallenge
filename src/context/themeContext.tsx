import React from 'react';
import { useColorScheme } from 'react-native';

import { ITheme, LightTheme, DarkTheme } from '../styles/theming';

import { createCtx } from './createCtx';

// Our context provider will provide this object shape
type TThemeContextState = {
  theme: ITheme;
  isDarkMode: boolean;
};

type TThemeContext = TThemeContextState & {};

const [useContext, ThemeContextProvider] =
  createCtx<TThemeContext>('themeContext');

interface Props {
  children?: React.ReactNode;
}

export const ThemeProvider = React.memo<Props>(({ children }) => {
  const isDarkMode = useColorScheme() === 'dark';

  const theme = isDarkMode ? DarkTheme : LightTheme;
  const [themeCtxState] = React.useState<TThemeContextState>({
    theme,
    isDarkMode,
  });

  const MemoizedValue = React.useMemo(() => {
    const value: TThemeContext = {
      ...themeCtxState,
    };
    return value;
  }, [themeCtxState]);

  return (
    <ThemeContextProvider value={MemoizedValue}>
      {children}
    </ThemeContextProvider>
  );
});

export const useTheme = useContext;
