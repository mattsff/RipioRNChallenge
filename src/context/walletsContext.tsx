import React, { useCallback, useEffect, useState } from 'react';

import { Fee } from '../types/fee';
import { ExchangeRates } from '../types/exchangeRates';
import { Wallet } from '../types/wallet';
import { bitcoin } from '../utils/mockData';
import { ripioApiProvider } from '../providers/apiProvider/ripioApiProvider';
import { feesApiProvider } from '../providers/apiProvider/feesApiProvider';

import { createCtx } from './createCtx';

interface WalletsContext extends WalletsContextState {
  getWallets: () => void;
  getExchangeRates: () => Promise<void>;
  getCommissions: () => Promise<void>;
  setWallets: (wallets: Wallet[]) => void;
  setInitialState: () => void;
}

interface WalletsContextState {
  wallets: Wallet[];
  exchangeRates?: ExchangeRates; // TODO: Support multiples currencies
  fees?: Fee; // TODO: Support multiples currencies
}

const INITIAL_STATE: WalletsContextState = {
  wallets: [],
  exchangeRates: undefined,
  fees: undefined,
};

const [useContext, WalletsContextProvider] =
  createCtx<WalletsContext>('walletsContext');

interface Props {
  children?: React.ReactNode;
}

export const WalletsProvider: React.FC<Props> = ({ children }) => {
  const [state, setState] = useState<WalletsContextState>(INITIAL_STATE);

  const getWallets = useCallback(() => {
    // Using mock data instead of getting this from network
    const walletBtc: Wallet = {
      currency: bitcoin,
      balance: 5,
      transactions: [],
    };
    setState((prevState) => ({ ...prevState, wallets: [walletBtc] }));
  }, []);

  const getExchangeRates = useCallback(async () => {
    const { data, error } = await ripioApiProvider.getRates();

    if (data && !error) {
      const exchangeRates: ExchangeRates = {
        sell: data.rates.ARSSELL,
        buy: data.rates.ARSBUY,
      };
      setState((prevState) => ({ ...prevState, exchangeRates: exchangeRates }));
    }
  }, []);

  const getCommissions = useCallback(async () => {
    const { data, error } = await feesApiProvider.getBTCFees();

    if (data && !error) {
      const fees: Fee = {
        network: 'BTC',
        amount: data.fastestFee,
      };
      setState((prevState) => ({ ...prevState, fees: fees }));
    }
  }, []);

  const setWallets = useCallback((wallets: Wallet[]) => {
    setState((prevState) => ({ ...prevState, wallets }));
  }, []);

  const setInitialState = useCallback(() => {
    setState(INITIAL_STATE);
  }, []);

  return (
    <WalletsContextProvider
      value={{
        wallets: state.wallets,
        exchangeRates: state.exchangeRates,
        fees: state.fees,
        getWallets,
        getExchangeRates,
        getCommissions,
        setWallets,
        setInitialState,
      }}>
      {children}
    </WalletsContextProvider>
  );
};

export const useWallets = useContext;
