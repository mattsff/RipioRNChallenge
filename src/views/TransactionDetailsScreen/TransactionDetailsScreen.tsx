import React from 'react';
import { ScrollView } from 'react-native-gesture-handler';
import { Header, Text } from '@rneui/themed';
import { View } from 'react-native';

import { useThemeAwareObject } from '../../hooks/useThemeAwareObject';
import { MainStackScreenProps } from '../../navigation/mainStack';
import { useTheme } from '../../context/themeContext';

import { createStyles } from './TransactionDetailsScreen.styles';
import { TransactionDetailsStrings } from './TransactionDetailsScreen.strings';
import { useTransactionDetails } from './TransactionDetailsScreen.hooks';

export const TransactionDetailsScreen: React.FC<
  MainStackScreenProps<'TransactionDetails'>
> = ({ route }) => {
  const transaction = route.params.transaction;
  const styles = useThemeAwareObject(createStyles);
  const { theme } = useTheme();

  const { transactionData, onGoBack } = useTransactionDetails(transaction);

  return (
    <View>
      <Header
        leftComponent={{
          icon: 'arrow-back',
          color: theme.onPrimary,
          onPress: onGoBack,
        }}
        centerComponent={{
          text: TransactionDetailsStrings.title,
          style: styles.headerTitle,
        }}
      />
      <ScrollView
        alwaysBounceVertical={false}
        showsVerticalScrollIndicator={false}>
        <View style={styles.container}>
          {transactionData.map((item, index) => {
            return (
              <View key={index}>
                <Text style={styles.label}>{item.label}</Text>
                <Text style={styles.value}>{item.value}</Text>
              </View>
            );
          })}
        </View>
      </ScrollView>
    </View>
  );
};
