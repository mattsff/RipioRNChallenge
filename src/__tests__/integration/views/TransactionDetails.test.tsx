import { NavigationContainer, RouteProp } from '@react-navigation/native';
import React from 'react';
import { render, screen } from '@testing-library/react-native';
import { ThemeProvider as ElementsThemeProvider } from '@rneui/themed';

import {
  MainStackNavigationProp,
  MainStackParamList,
} from '../../../navigation/mainStack';
import { WalletsProvider } from '../../../context/walletsContext';
import { TransactionDetailsScreen } from '../../../views/TransactionDetailsScreen/TransactionDetailsScreen';
import { ThemeProvider } from '../../../context/themeContext';
import { RNElementsTheme } from '../../../styles/theming';
import { mockTransaction } from '../../mockData';

const screenWithcustomNavigation = (
  navigate: jest.Mock<any, any>,
  route: RouteProp<MainStackParamList, 'TransactionDetails'>,
) =>
  render(
    <ElementsThemeProvider theme={RNElementsTheme}>
      <ThemeProvider>
        <NavigationContainer>
          <TransactionDetailsScreen
            navigation={
              {
                navigate,
              } as unknown as MainStackNavigationProp<'TransactionDetails'>
            }
            route={route}
          />{' '}
        </NavigationContainer>
      </ThemeProvider>
    </ElementsThemeProvider>,
    {
      wrapper: WalletsProvider,
    },
  );

describe('TransactionDetailsScreen', () => {
  test('snapshot', () => {
    const navigate = jest.fn();

    const route: RouteProp<MainStackParamList, 'TransactionDetails'> = {
      key: '',
      name: 'TransactionDetails',
      params: {
        transaction: mockTransaction,
      },
    };
    screenWithcustomNavigation(navigate, route);
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
