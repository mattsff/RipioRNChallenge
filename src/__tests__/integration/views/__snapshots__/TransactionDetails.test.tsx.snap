// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionDetailsScreen snapshot 1`] = `
[
  <View>
    <View
      replaceTheme={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#7908ff",
          "borderBottomColor": "#f2f2f2",
          "borderBottomWidth": 0.5,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingHorizontal": 10,
          "paddingVertical": 10,
        }
      }
      testID="headerContainer"
      updateTheme={[Function]}
    >
      <RNCSafeAreaView
        edges={
          [
            "left",
            "top",
            "right",
          ]
        }
        style={
          {
            "flexDirection": "row",
            "width": "100%",
          }
        }
      >
        <View
          style={
            {
              "alignItems": "flex-start",
              "flex": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "flex-start",
                "overflow": "hidden",
              }
            }
            testID="RNE__ICON__CONTAINER"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                {
                  "disabled": false,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              icon="arrow-back"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              testID="RNE__ICON__CONTAINER_ACTION"
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
                testID="RNE__ICON"
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    [
                      {
                        "color": "#FFFFFF",
                        "fontSize": 24,
                      },
                      {
                        "backgroundColor": "transparent",
                      },
                      {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      {},
                    ]
                  }
                  testID="RNE__ICON__Component"
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "flex": 3,
            }
          }
        >
          <Text
            numberOfLines={1}
            style={
              {
                "color": "#FFFFFF",
                "fontSize": 20,
                "fontStyle": "normal",
                "fontWeight": "bold",
                "letterSpacing": 0,
              }
            }
            text="Detalle de movimiento"
          >
            Detalle de movimiento
          </Text>
        </View>
        <View
          style={
            {
              "alignItems": "flex-end",
              "flex": 1,
            }
          }
        />
      </RNCSafeAreaView>
    </View>
    <RCTScrollView
      alwaysBounceVertical={false}
      collapsable={false}
      handlerTag={1}
      handlerType="NativeViewGestureHandler"
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      showsVerticalScrollIndicator={false}
      waitFor={
        [
          {
            "current": null,
          },
        ]
      }
    >
      <View>
        <View
          style={
            {
              "backgroundColor": "#FFFFFF",
              "borderRadius": 8,
              "marginHorizontal": 8,
              "marginVertical": 20,
              "padding": 16,
            }
          }
        >
          <View>
            <Text
              accessibilityRole="text"
              style={
                {
                  "color": "#6c6c6c",
                  "fontSize": 18,
                  "fontStyle": "normal",
                  "fontWeight": "600",
                  "letterSpacing": 0,
                  "marginTop": 20,
                }
              }
            >
              ID
            </Text>
            <Text
              accessibilityRole="text"
              style={
                {
                  "color": "#242424",
                  "fontSize": 16,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                  "letterSpacing": 0,
                  "marginTop": 4,
                }
              }
            >
              0
            </Text>
          </View>
          <View>
            <Text
              accessibilityRole="text"
              style={
                {
                  "color": "#6c6c6c",
                  "fontSize": 18,
                  "fontStyle": "normal",
                  "fontWeight": "600",
                  "letterSpacing": 0,
                  "marginTop": 20,
                }
              }
            >
              Fecha
            </Text>
            <Text
              accessibilityRole="text"
              style={
                {
                  "color": "#242424",
                  "fontSize": 16,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                  "letterSpacing": 0,
                  "marginTop": 4,
                }
              }
            >
              01/01/2022
            </Text>
          </View>
          <View>
            <Text
              accessibilityRole="text"
              style={
                {
                  "color": "#6c6c6c",
                  "fontSize": 18,
                  "fontStyle": "normal",
                  "fontWeight": "600",
                  "letterSpacing": 0,
                  "marginTop": 20,
                }
              }
            >
              Dirección
            </Text>
            <Text
              accessibilityRole="text"
              style={
                {
                  "color": "#242424",
                  "fontSize": 16,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                  "letterSpacing": 0,
                  "marginTop": 4,
                }
              }
            >
              3Mgwp1XyHNPC651JB1kzFUni8MziFk8Hi4
            </Text>
          </View>
          <View>
            <Text
              accessibilityRole="text"
              style={
                {
                  "color": "#6c6c6c",
                  "fontSize": 18,
                  "fontStyle": "normal",
                  "fontWeight": "600",
                  "letterSpacing": 0,
                  "marginTop": 20,
                }
              }
            >
              Monto
            </Text>
            <Text
              accessibilityRole="text"
              style={
                {
                  "color": "#242424",
                  "fontSize": 16,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                  "letterSpacing": 0,
                  "marginTop": 4,
                }
              }
            >
              5,0005 BTC
            </Text>
          </View>
          <View>
            <Text
              accessibilityRole="text"
              style={
                {
                  "color": "#6c6c6c",
                  "fontSize": 18,
                  "fontStyle": "normal",
                  "fontWeight": "600",
                  "letterSpacing": 0,
                  "marginTop": 20,
                }
              }
            >
              Costo de red
            </Text>
            <Text
              accessibilityRole="text"
              style={
                {
                  "color": "#242424",
                  "fontSize": 16,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                  "letterSpacing": 0,
                  "marginTop": 4,
                }
              }
            >
              0,00001 BTC
            </Text>
          </View>
          <View>
            <Text
              accessibilityRole="text"
              style={
                {
                  "color": "#6c6c6c",
                  "fontSize": 18,
                  "fontStyle": "normal",
                  "fontWeight": "600",
                  "letterSpacing": 0,
                  "marginTop": 20,
                }
              }
            >
              Tipo de movimiento
            </Text>
            <Text
              accessibilityRole="text"
              style={
                {
                  "color": "#242424",
                  "fontSize": 16,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                  "letterSpacing": 0,
                  "marginTop": 4,
                }
              }
            >
              RETIRO
            </Text>
          </View>
          <View>
            <Text
              accessibilityRole="text"
              style={
                {
                  "color": "#6c6c6c",
                  "fontSize": 18,
                  "fontStyle": "normal",
                  "fontWeight": "600",
                  "letterSpacing": 0,
                  "marginTop": 20,
                }
              }
            >
              Estado
            </Text>
            <Text
              accessibilityRole="text"
              style={
                {
                  "color": "#242424",
                  "fontSize": 16,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                  "letterSpacing": 0,
                  "marginTop": 4,
                }
              }
            >
              Completo
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>,
  " ",
]
`;
